contracts/SignatureValidatorManager.sol

```diff
@@ -0,0 +1,226 @@
+     *             0x24 to end - bytes containing signatures or signatureData either one of the below:
+     *             If first 4 bytes of signatureData are 0xb5c726cb i.e. bytes4(keccak256("Account712Signature(bytes32,bytes32,bytes)")); then it will be interpreted as follows:
+     *                  payload = abi.encodeWithSelector(0xb5c726cb, abi.encode(domainSeparator, structHash, signatures)
+     *                  Layout of `data` parameter in this case:
+     *                  0x00 to 0x04 - 4 bytes function selector for with the this contract is set as a function handler in the SafeProtocolManager i.e. 0x1626ba7e
+     *                  0x04 to 0x24 - 32 bytes hash of the signed message
+     *                  0x24 to 0x44 - 32 bytes offset to the start of `bytes` parameter
+     *                  0x44 to 0x64 - 32 bytes length of `bytes` parameter
+     *                  0x64 to 0x68 - 4 bytes of Signature selector
+     *                  0x68 to 0x88 - 32 bytes domain separator
+     *                  0x88 to 0xa8 - 32 bytes struct hash
+     *                  0xa8 to end - contains offset, length of bytes, and actual bytes containing signatures
+     *             Else:
+     *                 bytes containing signature data
+     *                 default validation flow will be used which will depend on the account implementation
+     *
+     */
+    function handle(address account, address sender, uint256 /* value */, bytes calldata data) external override returns (bytes memory) {
+        // Skip first 4 bytes of data as it contains function selector
+        (bytes32 messageHash, bytes memory signatureData) = abi.decode(data[0x4:], (bytes32, bytes));
+
+        address signatureValidatorHooksAddress = signatureValidatorHooks[account];
+        bytes memory prevalidationData;
+        bytes memory returnData;
+
+        if (signatureValidatorHooksAddress != address(0)) {
+            checkPermittedModule(signatureValidatorHooksAddress, MODULE_TYPE_SIGNATURE_VALIDATOR_HOOKS);
+            prevalidationData = ISafeProtocolSignatureValidatorHooks(signatureValidatorHooksAddress).preValidationHook(
+                account,
+                sender,
+                data
+            );
+        }
+
+        if (bytes4(data[0x64:0x68]) == SIGNATURE_VALIDATOR_SELECTOR) {
+            returnData = abi.encode(validateWithSignatureValdiator(account, sender, messageHash, data[0x68:]));
+
+            if (signatureValidatorHooksAddress != address(0)) {
+                ISafeProtocolSignatureValidatorHooks(signatureValidatorHooksAddress).postValidationHook(account, prevalidationData);
+            }
+            return returnData;
+        }
+
+        returnData = defaultValidator(account, messageHash, signatureData);
+        if (signatureValidatorHooksAddress != address(0)) {
+            ISafeProtocolSignatureValidatorHooks(signatureValidatorHooksAddress).postValidationHook(account, prevalidationData);
+        }
```

Comment:
control flow nit:


It seems like `if (signatureValidatorHooksAddress != address(0)) { ... }` always gets called right? So you could write this as:

```solidity
if (bytes4(data[0x64:0x68]) == SIGNATURE_VALIDATOR_SELECTOR) {
    returnData = abi.encode(validateWithSignatureValdiator(account, sender, messageHash, data[0x68:]));
} else {
    returnData = defaultValidator(account, messageHash, signatureData);
}

if (signatureValidatorHooksAddress != address(0)) {
    ISafeProtocolSignatureValidatorHooks(signatureValidatorHooksAddress).postValidationHook(account, prevalidationData);
}
```

Reply:
https://github.com/safe-global/safe-core-protocol/pull/109/commits/a2fb4d698ca417e00e375b1a556bd745bd608689